File: src/class13/deliverableexercise/hash/java/Course.java
package class13.deliverableexercise.hash.java;

import java.util.HashMap;
import java.util.Objects;

public class Course {
    private String name;
    private Integer courseCode;
    private TitularTeacher titularTeacher;
    private AssistantTeacher assistantTeacher;
    private Integer maxQuantityStudents;
    private HashMap<Integer, Student> studentHashMap = new HashMap<>();

    public Course(String name, Integer courseCode, Integer maxQuantityStudents) {
        this.name = name;
        this.courseCode = courseCode;
        this.maxQuantityStudents = maxQuantityStudents;
    }

    public Boolean addStudent(Student student) {
        if (studentHashMap.size() < maxQuantityStudents) {
            studentHashMap.put(student.getStudentCode(), student);
            return true;
        } else {
            return false;
        }
    }

    public void removeStudent(Student student) {
        studentHashMap.remove(student.getStudentCode());
        System.out.println("Aluno Removido do Curso " + name);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getCourseCode() {
        return courseCode;
    }

    public void setCourseCode(Integer courseCode) {
        this.courseCode = courseCode;
    }

    public TitularTeacher getTitularTeacher() {
        return titularTeacher;
    }

    public void setTitularTeacher(TitularTeacher titularTeacher) {
        this.titularTeacher = titularTeacher;
    }

    public AssistantTeacher getAssistantTeacher() {
        return assistantTeacher;
    }

    public void setAssistantTeacher(AssistantTeacher assistantTeacher) {
        this.assistantTeacher = assistantTeacher;
    }

    public Integer getMaxQuantityStudents() {
        return maxQuantityStudents;
    }

    public void setMaxQuantityStudents(Integer maxQuantityStudents) {
        this.maxQuantityStudents = maxQuantityStudents;
    }

    public HashMap<Integer, Student> getStudentHashMap() {
        return studentHashMap;
    }

    public void setStudentHashMap(HashMap<Integer, Student> studentHashMap) {
        this.studentHashMap = studentHashMap;
    }

    //Equals padrão que compara o objeto inteiro
//    @Override
//    public boolean equals(Object o) {
//        if (this == o) return true;
//        if (o == null || getClass() != o.getClass()) return false;
//        Course course = (Course) o;
//        return Objects.equals(name, course.name) &&
//                Objects.equals(courseCode, course.courseCode);
//    }

    //Equals que compara apenas o Código do Curso do objeto
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Course course = (Course) o;
        return Objects.equals(courseCode, course.courseCode);
    }
}

File: app/src/main/java/intellij/class13/deliverableexercise/hash/kotlin/Course.kt
package class13.deliverableexercise.hash.kotlin

import class13.deliverableexercise.list.kotlin.Course
import intellij.class13.deliverableexercise.hash.kotlin.TitularTeacher

class Course(var name: String, var courseCode: Int, var maxQuantityStudents: Int) {
    lateinit var titularTeacher: TitularTeacher
    lateinit var assistantTeacher: AssistantTeacher
    private var studentsHashMap = hashMapOf<Int, Student>()

    fun addStudent(student: Student): Boolean {
        return if (studentsHashMap.size < maxQuantityStudents) {
            studentsHashMap[student.studentCode] = student
            true
        } else {
            false
        }
    }

    fun removeStudent(student: Student) {
        studentsHashMap.remove(student.studentCode)
        println("Aluno Removido do Curso $name")
    }

    //Equals padrão que compara o objeto inteiro
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (javaClass != other?.javaClass) return false
//        other as Course
//        if (name != other.name) return false
//        if (courseCode != other.courseCode) return false
//        if (maxQuantityStudents != other.maxQuantityStudents) return false
//        return true
//    }

    //Equals que compara apenas o Código do Curso do objeto
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val course = other as Course
        return courseCode == course.courseCode
    }

}